Minio/DataModel/ServerSideEncryption.cs

```diff
@@ -0,0 +1,166 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Text;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+
+using Minio.Helper;
+
+namespace Minio.DataModel
+{
+  // Type of server side encryption
+   public enum EncryptionType {
+     SSE_C,
+     SSE_S3,
+     SSE_KMS
+   }
+
+   // ServerSideEncryption interface
+   public interface ServerSideEncryption
+   {
+    // GetType() needs to return the type of Server Side encryption
+    EncryptionType GetType() ;
+    // Marshals the server side encryption headers into dictionary
+    void Marshal(Dictionary<string,string> headers) ;
+   }
+
+   // Server side encryption with customer provided keys (SSE-C)
+    public class SSEC : ServerSideEncryption {
+    
+    // secret AES-256 Key
+    protected byte[] key;
+      public new EncryptionType GetType()
+      {
+        return EncryptionType.SSE_C;
+      }
+
+      public virtual void Marshal(Dictionary<string,string> headers) 
+      {
+        var md5SumStr = utils.getMD5SumStr(this.key);
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Algorithm", "AES256");
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Key",Convert.ToBase64String(this.key))  ;
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Key-Md5", md5SumStr);
+        return;
+      } 
+
+      public SSEC(byte[] key) 
+      {
+        if (key == null || key.Length != 32)
+        {
+          throw new ArgumentException("Secret key needs to be a 256 bit AES Key");
+        } 
+        this.key = key;
+      }
+    }
+
+  // Server side encryption option for source side SSE-C
+  // copy operation
+  public class SSECopy : SSEC 
+  {
+        public override void Marshal(Dictionary<string,string> headers) 
+        {
+          var md5SumStr = utils.getMD5SumStr(this.key);
+          headers.Add("X-Amz-Copy-Source-Server-Side-Encryption-Customer-Algorithm", "AES256");
+          headers.Add("X-Amz-Copy-Source-Server-Side-Encryption-Customer-Key",Convert.ToBase64String(key)) ;
+          headers.Add("X-Amz-Copy-Source-Server-Side-Encryption-Customer-Key-Md5", md5SumStr);
+          return;
+        }
+        public SSECopy(byte[] key) : base(key)
+        {
+        }
+
+        public SSEC CloneToSSEC()
+        {
+          return new SSEC(this.key);
+        }
+  }
+
+  // Server side encryption with S3 managed encryption keys (SSE-S3)
+  public class SSES3 : ServerSideEncryption
+  {
+      public new EncryptionType GetType()
+      {
+        return EncryptionType.SSE_S3;
+      }
+
+      public virtual void Marshal(Dictionary<string,string> headers)
+      {
+        headers.Add(Constants.SSEGenericHeader, "AES256"); 
+        return;
+      }
+  }
+
+  // Server side encryption with AWS KMS managed keys
+  public class SSEKMS : SSES3
```

Comment:
Why is this class derived from `SSES3`? Why can't this be derived from `ServerSideEncryption`?

Reply:
fixed